File: src/org/oopsc/statement/WhileStatement.java
package org.oopsc.statement;

import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

import org.oopsc.*;
import org.oopsc.expression.Expression;

/**
 * Die Klasse repräsentiert die Anweisung WHILE im Syntaxbaum.
 */
public class WhileStatement extends Statement {
	/** Die Bedingung der WHILE-Anweisung. */
	public Expression condition;

	/** Die Anweisungen im Schleifenrumpf. */
	public List<Statement> statements = new LinkedList<>();

	/**
	 * Konstruktor.
	 *
	 * @param condition
	 *        Die Bedingung der WHILE-Anweisung.
	 */
	public WhileStatement(Expression condition) {
		this.condition = condition;
	}

	public void defPass(SemanticAnalysis sem) throws CompileException {
		for (Statement s : this.statements) {
			s.defPass(sem);
		}
	}

	@Override
	public void refPass(SemanticAnalysis sem) throws CompileException {
		this.condition = this.condition.refPass(sem);
		this.condition.types = sem.types();
		this.condition.type.check(sem, sem.types().boolType(),
				this.condition.position);

		for (Statement s : this.statements) {
			s.refPass(sem);
		}
	}

	@Override
	public void print(TreeStream tree) {
		tree.println("WHILE");
		tree.indent();
		this.condition.print(tree);
		if (!this.statements.isEmpty()) {
			tree.println("DO");
			tree.indent();
			for (Statement s : this.statements) {
				s.print(tree);
			}
			tree.unindent();
		}
		tree.unindent();
	}

	@Override
	public void generateCode(CodeStream code, Stack<Context> contexts) {
		String whileLabel = code.nextLabel();
		String endLabel = code.nextLabel();

		code.println("; WHILE");
		code.println(whileLabel + ":");
		this.condition.generateCode(code, false);

		code.println("MRM R5, (R2) ; Bedingung vom Stapel nehmen");
		code.println("SUB R2, R1");
		code.println("ISZ R5, R5 ; Wenn 0, dann");
		code.println("JPC R5, " + endLabel + " ; Schleife verlassen");
		code.println("; DO");

		for (Statement s : this.statements) {
			s.generateCode(code, contexts);
		}

		code.println("; END WHILE");
		code.println("MRI R0, " + whileLabel);
		code.println(endLabel + ":");
	}
}

File: src/org/oopsc/statement/WhileStatement.scala
package org.oopsc.statement

import org.oopsc._
import org.oopsc.expression.{BooleanLiteralExpression, Expression}
import scala.collection.mutable.ListBuffer

class WhileStatement(var condition: Expression, var statements: ListBuffer[Statement]) extends Statement {
  override def refPass(sem: SemanticAnalysis) {
    this.condition.refPass(sem)
    this.condition.resolvedType.check(Types.boolType, this.condition.position)
    this.statements.foreach(_.refPass(sem))
  }

  override def optimPass() : Statement = {
    this.condition = this.condition.optimPass()

    this.condition match {
      case BooleanLiteralExpression(false, _) =>
        /* If the condition evaluates to false, return a NullStatement. */
        return new NullStatement

      case _ =>
    }

    this.statements = this.statements.map(_.optimPass())
    this
  }

  override def print(tree: TreeStream) {
    tree.println("WHILE")
    tree.indent
    this.condition.print(tree)

    if (!this.statements.isEmpty) {
      tree.println("DO")
      tree.indent
      this.statements.foreach(_.print(tree))
      tree.unindent
    }

    tree.unindent
  }

  override def generateCode(code: CodeStream, tryContexts: Int) {
    val whileLabel = code.nextLabel
    val endLabel = code.nextLabel

    code.println("; WHILE")
    code.println(s"$whileLabel:")

    this.condition match {
      case BooleanLiteralExpression(true, _) =>
        /* Minor optimisation: No need to generate evaluation code for the `true' literal. */
      case _ =>
        this.condition.generateCode(code, false)

        code.println("MRM R5, (R2) ; Take condition from the stack.")
        code.println("SUB R2, R1")
        code.println("ISZ R5, R5 ; If 0, then...")
        code.println(s"JPC R5, $endLabel ; ...leave the loop.")
    }

    code.println("; DO")

    for (s <- this.statements) {
      s.generateCode(code, tryContexts)
    }

    code.println(s"MRI R0, $whileLabel ; Another iteration.")
    code.println("; END WHILE")
    code.println(endLabel + ":")
  }
}